File: kakao/level2/수식_최대화.py
def solution(expression):
    operations = [('+','-','*'),('+','*','-'),('-','+','*'),('-','*','+'),('*','+','-'),('*','-','+')]
    answer = []
    for op in operations:
        a, b = op[0], op[1]
        tmp_list = []
        for e in expression.split(a):
            tmp = [f"({i})" for i in e.split(b)]
            tmp_list.append(f'({b.join(tmp)})')
        answer.append(abs(eval(a.join(tmp_list))))
    return max(answer)


expression = "100-200*300-500+20"
print(solution(expression)) # 60420

expression = "50*6-3*2"
print(solution(expression)) # 300

File: Java/programmers/kakao/level2/수식최대화.java
package CodeTest.Java.programmers.kakao.level2;

import java.util.*;

public class 수식최대화 {
    public long solution(String expression) {
        long answer = Long.MIN_VALUE;
        String[][] op = {{"+","-","*"},{"+","*","-"},{"-","*","+"},
                         {"-","+","*"},{"*","-","+"},{"*","+","-"}};
        List<String> arr = new ArrayList<>();
        
        int start = 0;
        for (int i=0; i<expression.length(); i++) {
            char x = expression.charAt(i);
            if (x == '-' || x == '+' || x == '*') {
                arr.add(expression.substring(start, i));
                arr.add(expression.substring(i, i+1));
                start = i+1;
            }
        }
        arr.add(expression.substring(start));
        
        for (int i=0; i<op.length; i++) {
            List<String> sub = new ArrayList<>(arr);
            for (int j=0; j<3; j++) {
                for (int k=0; k<sub.size(); k++) {
                    if (op[i][j].equals(sub.get(k))) {
                        sub.set(k-1, calc(sub.get(k-1), sub.get(k), sub.get(k+1)));
                        sub.remove(k); 
                        sub.remove(k);
                        k--;
                    }
                }
            }
            answer = Math.max(answer, Math.abs(Long.parseLong(sub.get(0))));
        }
        return answer;
    }
    
    private static String calc(String num1, String op, String num2) {
        long n1 = Long.parseLong(num1), n2 = Long.parseLong(num2);
        if (op.equals("+")) return String.valueOf(n1+n2);
        else if (op.equals("-")) return String.valueOf(n1-n2);
        else return String.valueOf(n1*n2);
    }
}
